src/main/java/scrolls/elder/model/DatastoreVersionStorage.java

```diff
@@ -0,0 +1,56 @@
+package scrolls.elder.model;
+
+import java.util.ArrayList;
+
+public class DatastoreVersionStorage {
+    private ArrayList<ReadOnlyDatastore> datastoreVersions;
+    private int pointer;
+
+    public DatastoreVersionStorage(ReadOnlyDatastore datastore) {
+        this.datastoreVersions = new ArrayList<ReadOnlyDatastore>();
+        datastoreVersions.add(datastore);
+        this.pointer = 0;
+    }
+
+    public boolean canUndo() {
+        // Pointer has at least 1 version of datastore to its left
+        if (pointer > 0) {
+            return true;
+        }
+
+        return false;
+    }
+
+    public boolean canRedo() {
+        int size = datastoreVersions.size() - 1;
+
+        // pointer is not at the end of the version history
+        if (pointer < size) {
+            return true;
+        }
+
+        return false;
+
+    }
+
+    public ReadOnlyDatastore executeUndo() {
+        assert this.canUndo() : "Undo command cannot be executed";
```

Comment:
Agreed. If I understand this correctly, the logic here will still hold, since `UndoCommand::Execute` has a guard clause checking `canUndo`, which throws an exception if it fails.  Hence, the assertion here is a supplementary sanity check to ensure correct usage of the `executeUndo` method 👍